test/jasmine/assets/domain_ref/components.js

```diff
@@ -0,0 +1,793 @@
+        'ARO position x:', JSON.stringify(aroposx), '\n',
+        'ARO position y:', JSON.stringify(aroposy), '\n',
+        'arrow axis pair:', arrowaxispair, '\n',
+        'xref:', xref, '\n',
+        'yref:', yref, '\n',
+    ].join(' ');
+}
+
+function testAnnotationCombo(combo, assert, gd) {
+    var axistypex = combo[0];
+    var axistypey = combo[1];
+    var axispair = combo[2];
+    var aroposx = combo[3];
+    var aroposy = combo[4];
+    var arrowaxispair = combo[5];
+    var xid = axispair[0];
+    var yid = axispair[1];
+    var xref = makeAxRef(xid, aroposx.ref);
+    var yref = makeAxRef(yid, aroposy.ref);
+    var axref = arrowaxispair[0] === 'p' ? 'pixel' : xref;
+    var ayref = arrowaxispair[1] === 'p' ? 'pixel' : yref;
+    var x0 = aroposx.value[0];
+    var y0 = aroposy.value[0];
+    var ax = axref === 'pixel' ? aroposx.pixel : aroposx.value[1];
+    var ay = ayref === 'pixel' ? aroposy.pixel : aroposy.value[1];
+    return Plotly.newPlot(gd, Lib.extendDeep({}, testMock))
+        .then(function(gd) {
+            return annotationTest(gd, {}, x0, y0, ax, ay, xref, yref, axref,
+                ayref, axistypex, axistypey, xid, yid);
+        }).then(function(testRet) {
+            assert(testRet);
+        });
+}
+
+// return a list of functions, each returning a promise that executes a
+// particular test. This function takes the keepGraphDiv argument, which if true
+// will prevent destroying the generated graph after the test is executed, and
+// an assert argument, which is a function that will be passed true if the test
+// passed.
+// {testCombos} is a list of combinations each of which will be passed to the
+// test function
+// {test} is the function returning a Promise that executes this test
+function comboTests(testCombos, test) {
+    var ret = testCombos.map(function(combo) {
+        return function(assert, gd) {
+            return test(combo, assert, gd);
+        };
+    });
+    return ret;
+}
+
+// return a list of strings, each describing a corresponding test
+// describe is a function taking a combination and returning a description of
+// the test
+function comboTestDescriptions(testCombos, desribe) {
+    var ret = testCombos.map(desribe);
+    return ret;
+}
+
+function annotationTestCombos() {
+    var testCombos = Array.from(iterable.cartesianProduct([
+        axisTypes, axisTypes, axisPairs, aroPositionsX, aroPositionsY, arrowAxis
+    ]));
+    testCombos = testCombos.map(
+        function(c, i) {
+            return c.concat(['graph-' + i]);
+        }
+    );
+    return testCombos;
+}
+
+function annotationTests() {
+    var testCombos = annotationTestCombos();
+    return comboTests(testCombos, testAnnotationCombo);
+}
+
+function annotationTestDescriptions() {
+    var testCombos = annotationTestCombos();
+    return comboTestDescriptions(testCombos, describeAnnotationComboTest);
+}
+
+
+function imageTestCombos() {
+    var testCombos = Array.from(iterable.cartesianProduct(
```

Comment:
Might be non-blocking:
`Array.from` is not supported in some browsers namely IE11: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from